Tests/Tests/STPCardTest.m

```diff
@@ -151,106 +142,176 @@ - (void)testStringFromFunding {
+- (void)testDescription {
+    STPCard *card = [STPCard decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:@"Card"]];
     XCTAssert(card.description);
 }
 
 #pragma mark - STPAPIResponseDecodable Tests
 
-- (NSDictionary *)completeAttributeDictionary {
-    return @{
-             @"id": @"1",
-             @"exp_month": @"12",
-             @"exp_year": @"2013",
-             @"funding": @"debit",
-             @"name": @"Smerlock Smolmes",
-             @"address_line1": @"221A Baker Street",
-             @"address_city": @"New York",
-             @"address_state": @"NY",
-             @"address_zip": @"12345",
-             @"address_country": @"USA",
-             @"last4": @"1234",
-             @"dynamic_last4": @"5678",
-             @"brand": @"MasterCard",
-             @"country": @"Japan",
-             @"currency": @"usd",
-             };
+- (void)testDecodedObjectFromAPIResponseRequiredFields {
+    NSArray<NSString *> *requiredFields = @[
+                                            @"id",
+                                            @"last4",
+                                            @"brand",
+                                            @"exp_month",
+                                            @"exp_year",
+                                            ];
+
+    for (NSString *field in requiredFields) {
+        NSMutableDictionary *response = [[STPTestUtils jsonNamed:@"Card"] mutableCopy];
+        [response removeObjectForKey:field];
+
+        XCTAssertNil([STPCard decodedObjectFromAPIResponse:response]);
+    }
+
+    XCTAssert([STPCard decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:@"Card"]]);
+}
+
+- (void)testDecodedObjectFromAPIResponseMapping {
+    NSDictionary *response = [STPTestUtils jsonNamed:@"Card"];
+    STPCard *card = [STPCard decodedObjectFromAPIResponse:response];
+
+    XCTAssertEqualObjects(card.cardId, @"card_103kbR2eZvKYlo2CDczLmw4K");
+    XCTAssertEqualObjects(card.addressCity, @"Pittsburgh");
+    XCTAssertEqualObjects(card.addressCountry, @"US");
+    XCTAssertEqualObjects(card.addressLine1, @"123 Fake St");
+    XCTAssertEqualObjects(card.addressLine2, @"Apt 1");
+    XCTAssertEqualObjects(card.addressState, @"PA");
+    XCTAssertEqualObjects(card.addressZip, @"19219");
+    XCTAssertEqual(card.brand, STPCardBrandVisa);
+    XCTAssertEqualObjects(card.country, @"US");
+    XCTAssertEqualObjects(card.currency, @"usd");
+    XCTAssertEqualObjects(card.dynamicLast4, @"5678");
+    XCTAssertEqual(card.expMonth, (NSUInteger)5);
+    XCTAssertEqual(card.expYear, (NSUInteger)2017);
+    XCTAssertEqual(card.funding, STPCardFundingTypeCredit);
+    XCTAssertEqualObjects(card.last4, @"4242");
+    XCTAssertEqualObjects(card.name, @"Jane Austen");
+
+    XCTAssertNotEqual(card.allResponseFields, response);
+    XCTAssertEqualObjects(card.allResponseFields, [response stp_dictionaryByRemovingNullsValidatingRequiredFields:@[]]);
```

Comment:
I'll come back to fixing this `stp_dictionaryByRemovingNullsValidatingRequiredFields` usage in another PR. The category is missing test coverage and we should probably deeply remove `[NSNull null]`s.